lib/types.nix

```diff
@@ -1029,8 +1029,26 @@ rec {
         getSubOptions = finalType.getSubOptions;
         getSubModules = finalType.getSubModules;
         substSubModules = m: coercedTo coercedType coerceFunc (finalType.substSubModules m);
-        typeMerge = t: null;
-        functor = (defaultFunctor name) // { wrapped = finalType; };
+        typeMerge = f':
+          let coercedType' = coercedType.typeMerge (f'.coercedType).functor;
```

Comment:
This should probably merge with `either` instead of a `typeMerge`, and support multiple coercion functions that way.

(not the original goal of this PR at all though...)